distributor: don't return 5xx on a single ingester outage

alanprot opened the pull request:
**What this PR does**:
Fixes cases where a single ingester outage cause 5XXs.

**Which issue(s) this PR fixes**:
Fixes https://github.com/cortexproject/cortex/issues/4381

**Checklist**
- [X] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

alanprot commented:
Just leaving this as a draft now to see if make sense...

@pracucci @bboreham @pstibrany @tomwilkie any thoughts?

Thanks

bboreham commented on pkg/ring/batch.go:
```diff
@@ -118,15 +134,28 @@ func (b *batchTracker) record(sampleTrackers []*itemTracker, err error) {
 	// The use of atomic increments here guarantees only a single sendSamples
 	// goroutine will write to either channel.
 	for i := range sampleTrackers {
+		verifyIfShouldReturn := func() {
+			if sampleTrackers[i].remaining.Dec() == 0 {
```
I would pass `i` as a parameter, to avoid any problem with closure binding.

bboreham commented on pkg/ring/batch.go:
```diff
@@ -118,15 +134,28 @@ func (b *batchTracker) record(sampleTrackers []*itemTracker, err error) {
 			}
 		} else {
 			if sampleTrackers[i].succeeded.Inc() != int32(sampleTrackers[i].minSuccess) {
+				verifyIfShouldReturn()
```
This can call `b.rpcsFailed.Inc()`.  Why are we incrementing the `Failed` number on the no-error branch?

alanprot replied:
We only do that when there is no remaining ingesters to call anymore - So even though the last ingester succeeded, we need to return an error. (in other words, because there is cases where we return an error even though the ingesters call succeeded).

lets consider the following scenario:

First Request:  4xx
Second Request: 5xx (no quorum yet)
Third Request: 2xx (in this case we did not reach maxFailures for 4xx nor 5xx)

In this particular scenario we can see that even though we have 2 errors (4xx + 5xx) we did not returned any error yet (as we had no quorum). What the `verifyIfShouldReturn` is doing is basically checking that there is no remaining calls to do for this particular sample and id not, return an error (pick any of the 4xx or 5xx). Its important to note that this function is called only if we did not reach the `minSuccess` needed to return a success.

Ps:
`b.rpcsFailed.Inc()` is here only to make sure we only return one error for the whole batch.

bboreham commented on pkg/distributor/distributor_test.go:
```diff
@@ -491,6 +494,106 @@ func TestDistributor_PushIngestionRateLimiter(t *testing.T) {
+
+	// Using 489 just to make sure we are not hitting the &limits
+	// Simulating 2 4xx and 1 5xx -> Should return 4xx
+	ingesters[0].failResp.Store(httpgrpc.Errorf(489, "Throttling"))
```
Is this a typo for 429, or am I missing some context?

alanprot replied:
its a typo :D

bboreham commented on pkg/distributor/distributor_test.go:
```diff
@@ -491,6 +494,106 @@ func TestDistributor_PushIngestionRateLimiter(t *testing.T) {
+	ingesters[1].failResp.Store(httpgrpc.Errorf(500, "InternalServerError"))
+	ingesters[2].failResp.Store(httpgrpc.Errorf(489, "Throttling"))
+
+	for i := 0; i < 1000; i++ {
```
Why 1000?

alanprot replied:
I tried many different implementations and many of them failed due race conditions... so this tests is testing multiples times just to make sure we dont have a race condition problem

bboreham commented on pkg/distributor/distributor_test.go:
```diff
@@ -491,6 +494,106 @@ func TestDistributor_PushIngestionRateLimiter(t *testing.T) {
+	ingesters[1].failResp.Store(httpgrpc.Errorf(489, "Throttling"))
+	ingesters[2].failResp.Store(httpgrpc.Errorf(500, "InternalServerError"))
+
+	for i := 0; i < 10000; i++ {
```
Why 10000?

alanprot replied:
Ditto

bboreham commented on pkg/distributor/distributor_test.go:
```diff
@@ -491,6 +494,106 @@ func TestDistributor_PushIngestionRateLimiter(t *testing.T) {
+	require.NoError(t, err)
+
+	// Give time to the ring get updated with the KV value
+	time.Sleep(5 * time.Second)
```
This is very long compared to sleeps in other tests -  is there another way to ensure propagation?

alanprot replied:
Yeah.. probably i can...

bboreham commented on pkg/distributor/distributor_test.go:
```diff
@@ -491,6 +494,106 @@ func TestDistributor_PushIngestionRateLimiter(t *testing.T) {
+	ingesters[1].happy.Store(true)
+	ingesters[2].happy.Store(true)
+
+	err := r.KVClient.CAS(context.Background(), ring.IngesterRingKey, func(in interface{}) (interface{}, bool, error) {
```
Why are we only doing a CAS on one ingester?

alanprot replied:
CAS return the whole ring here (in this case that has 3 ingesters):

```
r := in.(*ring.Desc)
```

And im just changing the state of ingester2 to "LEFT" -> So after this operation the ring will still have 3 ingester but ingester2 will be in left state.

bboreham commented on pkg/ring/batch.go:
```diff
@@ -118,15 +134,28 @@ func (b *batchTracker) record(sampleTrackers []*itemTracker, err error) {
 	// The use of atomic increments here guarantees only a single sendSamples
 	// goroutine will write to either channel.
 	for i := range sampleTrackers {
+		verifyIfShouldReturn := func() {
```
`verify` to me suggests a function without side-effects; this one has three.

alanprot replied:
yeah..
i agree that this logic is hard to understand.. lemme try to rewrite in another way.

bboreham reviewed (commented):
I couldn't really follow the central logic.  I suspect it does work, but it does so because of an intricate interplay. Can the code explain itself better or be less intricate?

alanprot commented:
I converted to draft as seems that the `batch.go` file was moved to another package. I just rewrote the solution in maybe a clearer way to see the feedbacks. If this is the way to go i can do the change in the `dskit` package.

alanprot commented:
Hi...

I was going to do this change in the DSKit but as this code is back in cortex I rebased it here (I can still do there if someone think it is worth).

@bboreham do you think with the comments now the code is a little more clear?

bboreham commented on pkg/ring/batch.go:
```diff
@@ -123,18 +139,34 @@ func (b *batchTracker) record(sampleTrackers []*itemTracker, err error) {
-				b.err <- err
+			// We count the error by error family (4xx and 5xx)
+			errCount := sampleTrackers[i].recordError(err)
+			// We should return an error if we reach the maxFailure (quorum) on a give error family OR
```
s/give/given/

alanprot replied:
Updated! :D

bboreham reviewed (commented):
Yes it does look better.  I am down to complaining about comments now :-)

Take this comment, which you didn't change:
```
	// If we succeed, decrement each sample's pending count by one. 
```
I think this sentence should have been removed long ago.  The following sentence is more true:

```
           If we reach
	// the required number of successful puts on this sample, then decrement the
	// number of pending samples by one.
```
And it is still true, but you have added a similar comment on line 155. Maybe we don't need this one any more.

```
  If we successfully push all samples to
	// min success instances, wake up the waiting rpc so it can return early.
```
This is still true; perhaps could move to line 155.
```
	// Similarly, track the number of errors, and if it exceeds maxFailures
	// shortcut the waiting rpc.
```
This bit is more complicated now: the comment on line 144 is closer.

```
	// The use of atomic increments here guarantees only a single sendSamples
	// goroutine will write to either channel.
```
Still true, but you now have more atomics so perhaps clarify it is specifically `rpcsFailed` and `rpcsPending` that are referred to.

Separately, in the test I would like a comment explaining the 1000 and 10000 "magic numbers".

bboreham reviewed (approved):
Approved!

alanprot commented:
Thanks a lot @bboreham  :D
